src/core/stdcpp/vector.d

```diff
@@ -0,0 +1,248 @@
+// std::vector declaration.
+//
+// Current caveats :
+// - mangling issues still exist
+// - won't work with custom allocators
+// - missing noexcept
+// - iterators are implemented as pointers
+// - no reverse_iterator nor rbegin/rend
+// - nothrow @trusted @nogc for most functions depend on knowledge
+//   of T's construction/destruction/assignment semantics
+///////////////////////////////////////////////////////////////////////////////
+
+import core.stdcpp.allocator;
+
+enum DefaultConstruct { value }
+
+/// Constructor argument for default construction
+enum Default = DefaultConstruct();
+
+extern(C++, "std"):
+
+extern(C++, class) struct vector(T, Alloc = allocator!T)
+{
+    static assert(!is(T == bool), "vector!bool not supported!");
+
+    alias value_type = T;
+    alias allocator_type = Alloc;
+    alias pointer = T*;
+    alias const_pointer = const(T)*;
+    alias difference_type = ptrdiff_t;
+    alias size_type = size_t;
+
+//    alias iterator = pointer;
+//    alias const_iterator = const_pointer;
+//    alias reverse_iterator
+//    alias const_reverse_iterator
+
+    alias as_array this;
+
+    // ctor/dtor
+    this(size_type count);
+    this(size_type count, ref const(value_type) val);
+    this(size_type count, ref const(value_type) val, ref const(allocator_type) al);
+    this(ref const(vector) x);
+//    this(iterator first, iterator last);
+//    this(iterator first, iterator last, ref const(allocator_type) al = defaultAlloc);
+//    this(const_iterator first, const_iterator last);
+//    this(const_iterator first, const_iterator last, ref const(allocator_type) al = defaultAlloc);
+//    extern(D) this(T[] arr)                                                     { this(arr.ptr, arr.ptr + arr.length); }
+//    extern(D) this(T[] arr, ref const(allocator_type) al = defaultAlloc)        { this(arr.ptr, arr.ptr + arr.length); }
+//    extern(D) this(const(T)[] arr)                                              { this(arr.ptr, arr.ptr + arr.length); }
+//    extern(D) this(const(T)[] arr, ref const(allocator_type) al = defaultAlloc) { this(arr.ptr, arr.ptr + arr.length); }
+    ~this();
+
+    ///
+    alias length = size;
+    ///
+    alias opDollar = length;
+
+    ref vector opAssign(ref const(vector) s);
+
+    // Iterators
+//    iterator begin() @trusted @nogc;
+//    const_iterator begin() const @trusted @nogc;
+//    const_iterator cbegin() const @trusted @nogc;
+//    iterator end() @trusted @nogc;
+//    const_iterator end() const @trusted @nogc;
+//    const_iterator cend() const @trusted @nogc;
+
+    // no reverse iterator for now.
+
+    // MSVC allocates on default initialisation in debug, which can't be modelled by D `struct`
+    @disable this();
+
+    // Capacity
+    size_type max_size() const nothrow @trusted @nogc;
+
+    void clear() nothrow;
+    void resize(size_type n);
+    void resize(size_type n, T c);
+    void reserve(size_type n = 0) @trusted @nogc;
```

Comment:
No point on putting `@trusted` on a definition. It's either `@safe` or `@system`